### ISSUE TITLE
allow LDAP authentication for 2.0

### ISSUE THREAD

**Issue by nathandunn** [GMOD/Apollo#115]
Shiro has a way to use LDAP:

http://isis.apache.org/components/security/shiro/using-ldap.html

and spring security has a direct plugin that supports both SHIRO and LDAP

http://grails.org/plugin/spring-security-shiro
http://grails.org/plugin/spring-security-ldap

all from:
http://grails.org/plugin/spring-security-core

**Comment by erasche** [comment]
Just wanted to chime in, this should be supported under [RemoteUserAuth](https://github.com/GMOD/Apollo/blob/master/src/main/java/org/bbop/apollo/web/user/remoteuser/RemoteUserAuthentication.java) with an upstream proxy like apache or nginx that support LDAP. Might save you the trouble of implementing it in java.

**Comment by monicacecilia** [comment]
Just waiting for someone to request this specifically.

**Comment by wieckr** [comment]
Well, if you were waiting for someone to chime in to request ldap auth, VectorBase would love it.

**Comment by childers** [comment]
We've been discussing this with @nathandunn for a while now too. I would like to put a vote in for a solution that supports [CAS](https://en.wikipedia.org/wiki/Central_Authentication_Service), or the CAS protocols.

**Comment by nathandunn** [comment]
@childers 

Looks like CAS is provided as well: http://grails.org/plugin/spring-security-cas

**Comment by nathandunn** [comment]
@wieckr  / @childers   I think that this will be a major focus of 2.0.4 as it also supports #978 as well.   We have to switch our security system over and then we can use those plugins directly.

**Comment by childers** [comment]
Yeah, CAS is one of the more widely used protocols, and a lot of frameworks support it. I hope it won't be too difficult to wire it into Apollo.

**Comment by nathandunn** [comment]
It shouldn’t be too difficult once we move the security framework back. 

Nathan

**Comment by erasche** [comment]
Just 2¢ from me. 

REMOTE_USER is essentially *always* easier to implement than specific auth frameworks if you don't need any of the features (e.g. group management/syncing from LDAP) and can live with the issue that HTTP auth doesn't support logout without ugly hacks. This is what galaxy used for years as the primary way to do LDAP, CAS support for varying site deployments. Galaxy only recently added support for "native" LDAP support.

All that's required for remote_user is to read a specific HTTP header and trust that whatever the upstream server sent is correct. Then admins can configure LDAP and CAS (or any other apache/nginx auth plugins) as they prefer.

**Comment by nathandunn** [comment]
@erasche  You make a compelling argument. I moved that code to a [1.0 branch for reference](https://github.com/GMOD/Apollo/blob/1.0/src/main/java/org/bbop/apollo/web/user/remoteuser/RemoteUserAuthentication.java
). 

Not too bad to write this code!   Is the idea behind this that you put Apollo behind Apache or NginX to and have them be responsible for the authentication with the upstream server and use a rewrite rule for Apollo if authenticated, populating the REMOTE_USER? 

https://wiki.galaxyproject.org/Admin/Config/ApacheExternalUserAuth

Just curious what @wieckr  and @childers  think.  

I think that Shiro supports something similar to this: 

http://www.programcreek.com/java-api-examples/index.php?api=org.apache.shiro.authc.AuthenticationException

**Comment by erasche** [comment]
@nathandunn yep, precisely. Upstream proxies like apache and nginx have wide support for dozens of auth methods and we have some assurance they're safely/correctly implemented. I just like passing responsibility for security to other projects. :)

**Comment by wieckr** [comment]
Using REMOTE_USER through apache rewrite rules would be fine for my purposes.  Ideally there would be some method of logging out, however I would take any centralized auth method over our currently segregated account management.

**Comment by nathandunn** [comment]
Probably the easiest thing to do in light of #1167 would be to add a simple Java Integration and stub passwords, so long as LDAP is the only authenticator.   Obviously you can still use REMOTE_USER, but adding the LDAP authentication would not be hard.   However, I don't really know enough about how this would work for an individual system to work with it.  Using individual Java (do people use Hashtable's anymore?) from [here](http://docs.oracle.com/javase/jndi/tutorial/ldap/security/ldap.html):

```
// Set up the environment for creating the initial context
Hashtable env = new Hashtable();
env.put(Context.INITIAL_CONTEXT_FACTORY, 
    "com.sun.jndi.ldap.LdapCtxFactory");
env.put(Context.PROVIDER_URL, "ldap://localhost:389/o=JNDITutorial");

// Authenticate as S. User and password "mysecret"
env.put(Context.SECURITY_AUTHENTICATION, "simple");
env.put(Context.SECURITY_PRINCIPAL, "cn=S. User, ou=NewHires, o=JNDITutorial");
env.put(Context.SECURITY_CREDENTIALS, "mysecret");

// Create the initial context
DirContext ctx = new InitialDirContext(env);
```

@wieckr  If you have any thoughts / comments on this I might be able to get you started on a PR.
